src/main/java/org/squiddev/cobalt/lib/StringPacker.java

```diff
@@ -0,0 +1,577 @@
+					pos += size;
+					break;
+				}
+				case 'x': {
+					pos++;
+					break;
+				}
+				case 'X': {
+					if (i + 1 >= fmt.length) throw new LuaError("invalid next option for option 'X'");
+					int size = 0;
+					if (fmt.bytes[++i] == 'i' || fmt.bytes[i] == 'I') {
+						while (i + 1 < fmt.length && fmt.bytes[i + 1] >= '0' && fmt.bytes[i + 1] <= '9') {
+							if (size >= Integer.MAX_VALUE / 10) throw new LuaError("bad argument #1 to 'packsize' (invalid format)");
+							size = (size * 10) + (fmt.bytes[++i] - '0');
+						}
+						if (size > 16 || size == 0)
+							throw new LuaError(String.format("integral size (%d) out of limits [1,16]", size));
+					} else size = packoptsize(fmt.bytes[i], 0);
+					if (size < 1) throw new LuaError("invalid next option for option 'X'");
+					if (pos % Math.min(size, alignment) != 0 && alignment > 1) for (int j = 0; pos % Math.min(size, alignment) != 0 && j < alignment; j++) pos++;
+					break;
+				}
+				case ' ': case '<': case '>': case '=': break;
+				case 's': case 'z': throw new LuaError("bad argument #1 to 'packsize' (variable-length format)");
+				default: throw new LuaError(String.format("invalid format option '%c'", fmt.bytes[i]));
+			}
+		}
+		return pos;
+	}
+
+	private static class UnpackIntState {
+		long result = 0;
+		int size = 0;
+	}
+
+	private static UnpackIntState unpackint(byte[] str, int offset, int size, ByteOrder endianness, int alignment, boolean signed) {
+		UnpackIntState retval = new UnpackIntState();
+		if (offset % Math.min(size, alignment) != 0 && alignment > 1) for (int i = 0; offset % Math.min(size, alignment) != 0 && i < alignment; i++) {offset++; retval.size++;}
+		for (int i = 0; i < size; i++) {
+			retval.result |= (((long) str[offset + i] & 0xFF) << ((endianness == ByteOrder.BIG_ENDIAN ? size - i - 1 : i) * 8));
+			retval.size++;
+		}
+		if (signed && (retval.result & (1 << (size * 8 - 1))) != 0)
+			for (int i = size; i < 8; i++) retval.result |= ((long)0xFF & 0xFF) << (i * 8);
+		return retval;
+	}
+
+	/**
+	 * string.unpack (fmt, s [, pos])
+	 *
+	 * Returns the values packed in string s (see string.pack) according to the format string fmt.
+	 * An optional pos marks where to start reading in s (default is 1).
+	 * After the read values, this function also returns the index of the first unread byte in s.
+	 */
+	static Varargs unpack(Varargs args) throws LuaError {
+		LuaString fmt = args.arg(1).checkLuaString();
+		LuaString str = args.arg(2).checkLuaString();
+		int pos = args.arg(3).optInteger(1);
+		if (pos < 0) pos = str.length + pos;
+		else if (pos == 0) throw new LuaError("bad argument #3 to 'unpack' (initial position out of string)");
+		else pos--;
+		if (pos > str.length || pos < 0) throw new LuaError("bad argument #3 to 'unpack' (initial position out of string)");
+		ByteOrder endianness = ByteOrder.nativeOrder();
+		int alignment = 1;
+		LuaTable retval = new LuaTable();
```

Comment:
It feels like it's probably worth using `List<LuaValue>` rather than `LuaTable`?